### PR TITLE
chain: DataSource Keeper

### LINKED ISSUE #491: Implement keeper and type for data sources
A data source is like a piece of reusable data sources that anyone can query. It can be referenced by an incremental ID, which can be used as a reference. This ticket contains the following work.
- Define key for `DataSourceCountStoreKey` and `DataSourceStoreKeyPrefix`
- Define data source struct, which should contains `Name`, `Fee`, `Executable`, and `Owner`.
- Implement the proper code in `/internal/types.go` and `/internal/keeper.go`

### PR DESCRIPTION
**Pull request by evilpeach**
Fixes #491

### CODE REVIEW

File: chain/x/zoracle/internal/types/key.go [hunk 1]
```diff
@@ -20,6 +20,9 @@ var (
 	// PendingListStoreKey is a key that help getting pending request
 	PendingListStoreKey = append(GlobalStoreKeyPrefix, []byte("PendingList")...)
 
+	// DataSourceCountStoreKey is a key that help getting to current data source count state variable
```
Comment (sorawit) [inline chain/x/zoracle/internal/types/key.go]: is the key that keeps the current data source count state variable

File: chain/x/zoracle/internal/types/util.go [hunk 2]
```diff
@@ -14,3 +14,13 @@ func uint64ToBytes(num uint64) []byte {
 	}
 	return buf.Bytes()
 }
+
+func int64ToBytes(num int64) []byte {
+	buf := new(bytes.Buffer)
+	err := binary.Write(buf, binary.BigEndian, num)
+	// Encode int64 must not have error
```
Comment (sorawit) [inline chain/x/zoracle/internal/types/util.go]: Another way to do it, according to how Cosmos-SDK does this, is:
```
        result := make([]byte, 8)
        binary.BigEndian.PutUint64(result, uint64(num))
        return result
```

That way, we don't need to import `bytes` and don't need to deal with error.

File: chain/x/zoracle/internal/keeper/keeper.go [hunk 3]
```diff
@@ -50,3 +50,29 @@ func (k Keeper) GetNextRequestID(ctx sdk.Context) uint64 {
...
+	}
+	err := k.cdc.UnmarshalBinaryLengthPrefixed(bz, &dataSourceNumber)
+	if err != nil {
+		panic(err)
+	}
+	return dataSourceNumber
+}
+
+// GetNextDataSourceID increments and returns the current number of data source.
+// If the global data source count is not set, it initializes it with value 0.
```
Comment (sorawit) [inline chain/x/zoracle/internal/keeper/keeper.go]: it initializes the value and returns 1

File: chain/x/zoracle/internal/types/data_source.go [hunk 4]
```diff
@@ -0,0 +1,23 @@
+package types
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// DataSource is a type to store detail of data source
+type DataSource struct {
```
Comment (sorawit) [inline chain/x/zoracle/internal/types/data_source.go:8]: nit: fullstop.

File: chain/x/zoracle/internal/types/data_source.go [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+	Owner      sdk.AccAddress `json:"owner"`
+	Name       string         `json:"name"`
+	Fee        sdk.Coins      `json:"fee"`
+	Executable []byte         `json:"executable"`
+}
+
+// NewDataSource - contructor of DataSource struct
```
Comment (sorawit) [inline chain/x/zoracle/internal/types/data_source.go]: Let's follow the pattern that we usually do. NewDataSource creates a new DataSource instance.

File: chain/x/zoracle/internal/keeper/keeper.go [hunk 6]
```diff
@@ -50,3 +50,29 @@ func (k Keeper) GetNextRequestID(ctx sdk.Context) uint64 {
 	store.Set(types.RequestsCountStoreKey, bz)
 	return requestNumber + 1
 }
+
+// GetDataSourceCount returns the current number of all data sources ever exist.
+func (k Keeper) GetDataSourceCount(ctx sdk.Context) int64 {
+	var dataSourceNumber int64
```
Comment (sorawit) [inline chain/x/zoracle/internal/keeper/keeper.go]: Why don't we name this `dataSourceCount`? ditto below.

File: chain/x/zoracle/internal/keeper/data_source.go [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+		return types.DataSource{}, types.ErrRequestNotFound(types.DefaultCodespace)
+	}
+
+	bz := store.Get(types.DataSourceStoreKey(id))
+	var dataSource types.DataSource
+	k.cdc.MustUnmarshalBinaryBare(bz, &dataSource)
+	return dataSource, nil
+}
+
+// CheckDataSourceExists checks if the data source of this id either valid on store or not.
```
Comment (sorawit) [inline chain/x/zoracle/internal/keeper/data_source.go]: // CheckDataSourceExists checks if the data source of this ID exists in the storage.

File: chain/x/zoracle/internal/keeper/data_source.go [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+)
+
+// SetDataSource is a function to save data source to the given ID.
+func (k Keeper) SetDataSource(ctx sdk.Context, id int64, dataSource types.DataSource) {
+	store := ctx.KVStore(k.storeKey)
+	store.Set(types.DataSourceStoreKey(id), k.cdc.MustMarshalBinaryBare(dataSource))
+}
+
+// GetDataSource returns the entire Request metadata struct.
+func (k Keeper) GetDataSource(ctx sdk.Context, id int64) (types.DataSource, sdk.Error) {
```
Comment (sorawit) [inline chain/x/zoracle/internal/keeper/data_source.go:15]: returns the entire DataSource struct for the given ID

File: chain/x/zoracle/internal/keeper/data_source.go [hunk 9]
```diff
@@ -0,0 +1,32 @@
+package keeper
+
+import (
+	"github.com/bandprotocol/d3n/chain/x/zoracle/internal/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// SetDataSource is a function to save data source to the given ID.
+func (k Keeper) SetDataSource(ctx sdk.Context, id int64, dataSource types.DataSource) {
```
Comment (sorawit) [inline chain/x/zoracle/internal/keeper/data_source.go:9]: SetDataSource saves the given data source with the given ID to the storage.

**Review by sorawit — CHANGES_REQUESTED** [review]
Logic looks great. Just a few wording comments and this is ready to go!

**Review by sorawit — DISMISSED** [review]
LGTM

**Review by taobun — APPROVED** [review]
LGTM

**Review by sorawit — APPROVED** [review]

**Pull request merged**